### PR TITLE
use ruamel.yaml instead of pyyaml for comment roundtrip

### PR DESCRIPTION
**Pull request by msarahan**
This is a different parser that keeps comments in yaml files.  It is more robust than our hand-coded solution (which is removed by this commit).  The library is ruamel.yaml (https://pypi.python.org/pypi/ruamel.yaml, fork of pyyaml), which has also replaced pyyaml in the cookiecutter project (https://github.com/audreyr/cookiecutter/pull/558)

This obviates the --force option.  For legacy reasons, I left it in.

There is one caveat to the ruamel.yaml library: it does not handle whitespace well.  It doesn't barf, it just doesn't do what you expect.  Namely, for lists, you'll end up with your list having the whitespace in between list items.  This is a known issue: https://bitbucket.org/ruamel/yaml/issues/16/keeping-newline-information#comment-25446142

Ping @groutr @kalefranz

### CODE REVIEW

**Comment by kalefranz** [comment]
This is awesome. Does ruamel require compiling or is it pure python?

File: .travis.yml [hunk 1]
```diff
@@ -32,6 +32,7 @@ install:
   - if [[ "$PYCOSAT" ]]; then
       conda install pycosat=$PYCOSAT;
     fi
+  - pip install ruamel.yaml
```
Comment (JanSchulz) [inline .travis.yml]: This is also in the conda-forge channel... :-)

Comment (msarahan) [inline .travis.yml]: Nice!  Thanks for pointing that out.  What projects are you aware of that use this?

Comment (JanSchulz) [inline .travis.yml]: At least conda-smithy does use it (https://github.com/conda-forge/conda-smithy/blob/master/conda_smithy/lint_recipe.py)...

**Comment by AvdN** [comment]
I updated the CommentedSeq class to support moving of comments on `.insert()` and `.pop()`.
in version 0.10.23 of ruamel.yaml.

The tests are [here](https://bitbucket.org/ruamel/yaml/src/90e4fb18d899b2a10de7138f77641ebbc0035399/_test/test_comments.py?at=default&fileviewer=file-view-default#test_comments.py-284). I think that should solve your issue, let me know if it doesn't.

Please not that comments hang off sequence items and any blank line after the last item in a sequence is part of that last comment. So inserting at the end, or appending, will result in the blank line being between the last item (just inserted) and the one before that. This is IMO not solvable without some better semantics as to when stuff (comments/blank lines) belongs to the previous (last scalar) parsed, the previous toplevel construct or the next construct, or its first item.

**Comment by kalefranz** [comment]
Hey @msarahan, how far are we from making this happen? Can you rebase against master to fix the merge conflicts and pull in the latest test configuration?

**Comment by msarahan** [comment]
I need to test this with spaces put back in the tests, then rebase as you say.  I will do so as soon as I can - probably by Monday morning.

I don't anticipate major issues.  However, this currently kills indentation, and is backwards incompatible.  Before you merge this, you either need to decide that we need to fix this (which could be a major project), or accept losing indentation (but improving our ability to parse with any indentation, and keep comments)

**Comment by AvdN** [comment]
@msarahan If it helps, I have a routine [here](https://github.com/pre-commit/pre-commit/pull/211#issuecomment-186466605) that tries to find your codes indentation level, that you can
feed into `ruamel.yaml.dump(......., indent=your_indent)`. That will still mangle inconsistent indenting because no information is kept on a individual basis, and also key aligns the dashes of a list.

You can of course also hard-code 4 (or any other positive value) indents if that is what always should be used instead of the default of 2 that ruamel.yaml uses.

**Comment by msarahan** [comment]
@AvdN thank you for following this issue.

I found the example at https://bitbucket.org/ruamel/yaml/issues/12/indent-incompletely-implemented and tried that.  The issue that I encountered was that subsequent dictionary keys below a list were being run into the last list item, effectively corrupting the read.

I will test the other example code you just posted soon - probably tomorrow - and let you know how it works for us.  Hard-coding indentation is just fine, I think, but people do not like the visual look of no indentation for list items.

Is the code at github now the authoritative source, or should we look at bitbucket for issues and such?

**Comment by AvdN** [comment]
@msarahan There were actually multiple issues the original PyYAML code is not willing to indent a scalar in a block sequence (which you only notice if you have non-scalars as well, and set `indent>2`). The scalar was always indented 2 resulting 
in `res = yaml.dump({"x": [{"b": 1}, 2]}, default_flow_style=False, indent=4)` giving

```
x:
-   b: 1
- 2
```

In [ruamel.yaml 0.11.3](https://pypi.python.org/pypi/ruamel.yaml#indentation-of-block-sequences) this is fixed and in addition there is the `block_seq_indent` keyword parameter with which you can push the dashes in a block_sequence inward. Thus 
`res = ruamel.yaml.dump({"x": [{"b": 1}, 2]}, default_flow_style=False, indent=4, block_seq_indent=2)`
gives you:

```
x:
  - b: 1
  - 2
```

I hope this gives you fine enough control. 

The authoritative source is on bitbucket, please file bugs there (including non-working example code). But in principle you should be able to install from PyPI, unless you want to dive into the gory details.

**Comment by msarahan** [comment]
@AvdN your updates worked flawlessly.  Thank you for your work, and especially for helping us make ruamel work for us.  The one remaining issue is that our code tests with "boolean" things values like "yes."  I get errors like: 

    Error: Key: changeps1; yes is not a YAML boolean.\n'
                 + Error: Key: changeps1; yes is not a YAML boolean.

This seems to be true of both "yes" and "Yes", as well as simply "Y"

The official YAML spec at http://www.yaml.org/refcard.html seems to support these.  Is there a reason why ruamel.yaml (or perhaps pyyaml) does not support it?

**Comment by AvdN** [comment]
@msarahan  TL;DR add  `version=”1.1”` to your load routine:  `round_trip_load(version="1.1")` (or `load(stream, Loader=ruamel.yaml.RoundTripLoader, version="1.1")`

The "official ref card" is the refcard for PyYAML which only supports YAML1.1 completely and that and some other features were dropped in the [YAML 1.2 specification](http://www.yaml.org/spec/1.2/spec.html) (back in 2009), which IMO you should follow. I recently made ruamel.yaml (already with better support for 1.2 than PyYAML) version aware [as described here](https://github.com/conda/conda/pull/2077#issuecomment-189766037) and announced  the top of the README that it defaults to 1.2. 

After loading (with version="1.1") this will be booleans internally undistinguishable from ones loaded from true or false, and on dumping these will all be dumped as true resp. false. It is probably possible to dump booleans as Yes/No with a little code grafting, but I am not a fan of that, especially as there are also people that are confused that you need to quote Yes, etc if you specify a scalar and want it to load as a string (using `load` or `safe_load`), which is an extra reasons to default to 1.2.
 
With `version=1.1` the YAML code in your tests seem to load fine: 
[conda_roundtrip.zip](https://github.com/conda/conda/files/149774/conda_roundtrip.zip)

**Comment by msarahan** [comment]
@AvdN thanks for your insight.  I would like to simply remove the YAML 1.1 stuff and keep up with the times, especially since that change was made so long ago.  However, I must err on the side of caution since our installed user base may already have a sizable collection of deprecated config files.

Thanks for the PR to add this fix.  @kalefranz, this is ready to go thanks to Anthon.  Travis builds are failing because conda-build is complaining that ruamel.yaml is not available as a conda package.  I can add that before or after you merge this.  It's already available on conda-forge, so perhaps this would make a good start towards using conda-forge for our stuff?

**Comment by kalefranz** [comment]
This is all totally F-ing awesome. @AvdN you've been so responsive and helpful. I've started questioning why yaml isn't already part of the python stdlib, and I'm finding everybody is wondering the same thing. A few of us at Continuum even want to start a PEP to make this happen. And your implementation is obviously the best. @msarahan, thanks for pushing this through. So awesome.

**Comment by kalefranz** [comment]
@msarahan Can you rebase against master? Will probably merge later today or tomorrow...

**Comment by kalefranz** [comment]
Woohoo!!!!!

**Comment by msarahan** [comment]
Thanks again @AvdN

**Comment by AvdN** [comment]
@msarahan, you're welcome and congrats. It is great to see something one spends quite some time on, being useful to others.
If/when I add a list of projects that use `ruamel.yaml` to the documentation, can I put `conda` on there?

**Comment by msarahan** [comment]
Sure - you've certainly earned it.

**Pull request merged**